src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,272 @@
+#include "ServicePool.h"
+
+#include <string.h>
+
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/RandUtils.h>
+
+using chip::Mdns::kMdnsTextMaxSize;
+using chip::Mdns::MdnsService;
+using chip::Mdns::TextEntry;
+
+namespace {
+
+// Only for MdnsService inside the ServicePool::Entry
+void FreeTextEntries(MdnsService * service)
```

Comment:
It turned out that it's not necessary to handle all these dynamic memories for the text here. According to the spec all the provisioned devices don't have text entries in their published services.  In case they publish some text in the future, we can save the information of interest rather than the full text. This will reduce both the code complexity and the resource consumption.